kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/auth/SpnegoAuthHeaderGenerator.java

```diff
@@ -55,45 +57,43 @@ public String generateAuthHeader() {
    * YarnClientUtils::generateToken
    *
    * @param server - hostname to contact
-   * @throws IOException
-   * @throws InterruptedException
+   * @throws Exception
    */
-  public static String generateToken(String server) throws IOException, InterruptedException {
-    UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
-    LOG.debug("The user credential is {}", currentUser);
-    String challenge =
-        currentUser.doAs(
-            new PrivilegedExceptionAction<String>() {
-              @Override
-              public String run() throws Exception {
-                try {
-                  GSSManager manager = GSSManager.getInstance();
-                  // GSS name for server
-                  GSSName serverName =
-                      manager.createName("HTTP@" + server, GSSName.NT_HOSTBASED_SERVICE);
-                  // Create a GSSContext for authentication with the service.
-                  // We're passing client credentials as null since we want them to
-                  // be read from the Subject.
-                  // We're passing Oid as null to use the default.
-                  GSSContext gssContext =
-                      manager.createContext(
-                          serverName.canonicalize(null), null, null, GSSContext.DEFAULT_LIFETIME);
-                  gssContext.requestMutualAuth(true);
-                  gssContext.requestCredDeleg(true);
-                  // Establish context
-                  byte[] inToken = new byte[0];
-                  byte[] outToken = gssContext.initSecContext(inToken, 0, inToken.length);
-                  gssContext.dispose();
-                  // Base64 encoded and stringified token for server
-                  LOG.debug("Got valid challenge for host {}", serverName);
-                  return new String(
-                      Base64.getEncoder().encode(outToken), StandardCharsets.US_ASCII);
-                } catch (GSSException e) {
-                  LOG.error("Error: ", e);
-                  throw new AuthenticationException("Failed to generate token", e);
-                }
-              }
-            });
-    return challenge;
+  private String generateToken(String server) throws Exception {
+    Subject subject;
+    try {
+      Class<?> ugiClz = Class.forName(UGI_CLASS);
+      Method ugiGetCurrentUserMethod = ugiClz.getDeclaredMethod("getCurrentUser");
+      Object ugiCurrentUser = ugiGetCurrentUserMethod.invoke(null);
+      LOG.debug("The user credential is {}", ugiCurrentUser);
+      Field ugiSubjectField = ugiCurrentUser.getClass().getDeclaredField("subject");
+      subject = (Subject) ugiSubjectField.get(ugiCurrentUser);
+    } catch (ClassNotFoundException e) {
+      // TODO do kerberos authentication using JDK class directly
```

Comment:
We actually can do Kerberos login using JDK classes directly, as long as the user provides principle and keytab.

Reply:
Some further thoughts: we can add `principle` and `keytabPath` in RestClientConf, and use them to do Kerberos authentication if the configuration is provided by the user, then fall back to using Hadoop UGI.
And we can rework the Kerberos auth in kyuubi jdbc driver in the same way.
cc @zhouyifan279

Reply:
hmm it's a big change, can we make a new pr for this ?

Reply:
It's a plan, the current change is a simple refactor.